llama.h

```diff
@@ -263,6 +289,9 @@ extern "C" {
     LLAMA_API void llama_sample_typical(struct llama_context * ctx, llama_token_data_array * candidates, float p, size_t min_keep);
     LLAMA_API void llama_sample_temperature(struct llama_context * ctx, llama_token_data_array * candidates, float temp);
 
+    /// @details Apply constraints from grammar
+    LLAMA_API void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * candidates, const struct llama_grammar * grammar);
```

Comment:
Can we make llama_grammar as a structure with two callbacks? So the other implementation of it can support context aware state machine instead?

Reply:
Do you mean like, the caller would provide the implementation of `llama_grammar` (via callbacks), from which the llama API determines which tokens are valid?

Reply:
yes, so llama code will not assume the grammar implementation.

Reply:
Yeah, I'm open to that idea, assuming the grammar interface itself generalizes well to other implementations. I kind of designed this with the specific implementation in mind so that's not a guarantee.